driver/pace/driver/run.py

```diff
@@ -279,11 +390,42 @@ def total_time(self) -> timedelta:
             days=self.days, hours=self.hours, minutes=self.minutes, seconds=self.seconds
         )
 
+    @staticmethod
+    def _set_kwargs_from_serialbox(kwargs: Dict[str, Any], config: SerialboxConfig):
+        kwargs["nx_tile"] = config.namelist.npx - 1
+        kwargs["nz"] = config.namelist.npz
+        kwargs["layout"] = tuple(config.namelist.layout)
+        kwargs["dt_atmos"] = float(config.namelist.dt_atmos)
+        kwargs["dycore_config"] = fv3core.DynamicalCoreConfig.from_f90nml(
+            config.f90_namelist
+        )
+        kwargs["physics_config"] = fv3gfs.physics.PhysicsConfig.from_f90nml(
+            config.f90_namelist
+        )
+        kwargs["days"] = config.namelist.days
+        kwargs["hours"] = config.namelist.hours
+        kwargs["minutes"] = config.namelist.minutes
+        kwargs["seconds"] = config.namelist.seconds
+        return kwargs
+
     @classmethod
     def from_dict(cls, kwargs: Dict[str, Any]) -> "DriverConfig":
-        kwargs["layout"] = tuple(kwargs["layout"])
         initialization_type = kwargs["initialization_type"]
-        if initialization_type == "baroclinic":
+        if initialization_type == "serialbox":
+            config = kwargs.get("initialization_config", {})
+            config.update(kwargs.get("stencil_config", {}))
+            kwargs["initialization_config"] = dacite.from_dict(
+                data_class=SerialboxConfig,
+                data=config,
+                config=dacite.Config(strict=True),
+            )
+            kwargs = cls._set_kwargs_from_serialbox(
+                kwargs, kwargs["initialization_config"]
+            )
+            return dacite.from_dict(
+                data_class=cls, data=kwargs, config=dacite.Config(strict=True)
+            )
```

Comment:
This return statement here makes it a lot harder to tell what this function does. There's code below that does necessary processing to integrate the top level driver configuration with values in DycoreConfig and PhysicsConfig. Either that code is duplicated somehow in `set_kwargs_from_serialbox`, which makes it harder to maintain that code (has to be updated in two places), or that code is not duplicated and the behavior of the rest of the configuration depends on how the initialization is set.

It also looks like right now the value of `initialization_type` makes the non-initialization part of the driver configuration yaml behave differently in ways I wouldn't expect as a user.

Is there any way this can be reduced to `initialization_class=SerialboxConfig` with all the concerns met by `set_kwargs_from_serialbox` treated automatically within the `SerialboxConfig` class? You have full control over the attributes of SerialboxConfig to make it work when you call `dacite.from_dict`, and you have access to a `__post_init__` dataclass method to add/modify arbitrary attributes afterwards.

Reply:
Okay updated per our discussion. It's still not super clean because I couldn't find a way to do this nicely without adding more inputs in the serialbox config yaml. More comments below.